lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp

```diff
@@ -558,6 +558,95 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
+        int64_t indicesLastDim = indicesShape.back();
+        // Given data tensor of rank r >= 1, indices tensor of rank q >= 1, and
+        // batch_dims integer b, onnx.gather_nd gathers slices of data into an
+        // output tensor of rank q + r - indices_shape[-1] - 1 - b.
+        // indices_shape[-1] must be static to have deterministic output rank.
+        if (indicesLastDim == Torch::kUnknownSize)
+          return failure();
+
+        // Reduce indices to properly map it to the flattened data.
+        SmallVector<int64_t> indexMultiplier;
+        int64_t prod = 1;
+        for (int64_t i = indicesLastDim - 1; i >= 0; i--) {
+          indexMultiplier.push_back(prod);
+          prod *= dataShape[i];
+        }
+        std::reverse(indexMultiplier.begin(), indexMultiplier.end());
+        Value indexMultiplierList =
+            createConstantIntList(binder, rewriter, indexMultiplier);
+        Value none = rewriter.create<Torch::ConstantNoneOp>(loc);
+        Value boolFalse = rewriter.create<Torch::ConstantBoolOp>(loc, false);
+        Type multiplierTy =
+            Torch::ValueTensorType::get(ctx, llvm::ArrayRef({indicesLastDim}),
+                                        indicesTy.getOptionalDtype());
+        Value multiplierTensor = rewriter.create<Torch::AtenTensorOp>(
+            loc, multiplierTy, indexMultiplierList,
+            Torch::getDtypeIntValueForType(rewriter, loc,
+                                           indicesTy.getOptionalDtype()),
+            none, boolFalse);
+        Value updatedIndices = rewriter.create<Torch::AtenMulTensorOp>(
+            loc, indicesTy, indices, multiplierTensor);
+        // Reduce the last dimension of indices.
+        indicesShape = indicesShape.drop_back();
+        Type reducedIndicesTy = Torch::ValueTensorType::get(
+            ctx, llvm::ArrayRef(indicesShape), indicesTy.getOptionalDtype());
+        Value dimList = createConstantIntList(
+            binder, rewriter, SmallVector<int64_t>({indicesRank - 1}));
+        Value reducedIndices = rewriter.create<Torch::AtenSumDimIntListOp>(
+            binder.getLoc(), reducedIndicesTy, updatedIndices, dimList,
+            boolFalse, none);
+
+        // Flatten first indices_shape[-1] dimensions of data.
+        SmallVector<int64_t> flattenDataShape({prod});
+        flattenDataShape.insert(flattenDataShape.end(),
+                                dataShape.begin() + indicesLastDim,
+                                dataShape.end());
+        auto flattenDataTy = Torch::ValueTensorType::get(
+            ctx, llvm::ArrayRef(flattenDataShape), dataTy.getOptionalDtype());
+        Value zero = rewriter.create<Torch::ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(0));
+        Value end = rewriter.create<Torch::ConstantIntOp>(
+            loc, rewriter.getI64IntegerAttr(indicesLastDim - 1));
+        Value flattenData = rewriter.create<Torch::AtenFlattenUsingIntsOp>(
```

Comment:
Generally it is bad to flatten your data tensor from a gather as for embedding cases (where gathers are a common use) this would be reshaping a massive tensor where only a subsection of the data is used. Is there an alternative that does not involve manipulating the `data` type?

Reply:
The best thing I got is to flatten only index_shape[-1] dims of data. Since it involves indexing at multiple axis, I don't see a way without flattening data at least that much. Thanks!
Lmk if there is a better way.